javascript/ql/src/semmle/javascript/dataflow/internal/InterProceduralTypeInference.qll

```diff
@@ -190,3 +127,119 @@ private class AnalyzedThisInPropertyFunction extends AnalyzedThisExpr {
   }
 }
 
+/**
+ * Holds if the execution of function `f` may complete normally without
+ * encountering a `return` or `throw` statement.
+ *
+ * Note that this is an overapproximation, that is, the predicate may hold
+ * of functions that cannot actually complete normally, since it does not
+ * account for `finally` blocks and does not check reachability.
+ */
+predicate mayReturnImplicitly(Function f) {
+  exists (ConcreteControlFlowNode final |
+    final.getContainer() = f and
+    final.isAFinalNode() and
+    not final instanceof ReturnStmt and
+    not final instanceof ThrowStmt
+  )
+}
+
+/**
+ * A call with inter-procedural type inference for the return value.
+ */
+abstract class CallWithAnalyzedReturnFlow extends DataFlow::CallNode, DataFlow::AnalyzedValueNode {
+
+  /**
+   * Gets a called function.
+   */
+  abstract Function getAFunction();
+  
+  /**
+   * Gets a return value for this call.
+   */
+  AbstractValue getAReturnValue() {
+    exists (Function f | f = getAFunction() |
+      if f.isGenerator() or f.isAsync() then
```

Comment:
This logic feels like it should go onto a class modelling functions, not the one modelling calls. How about introducing `AnalyzedFunction` or something like that?

Reply:
I have made the existing `AnalyzedFunction` public, it is now the only public class in `BasicExprTypeInference`. Is that OK?